libraries/AP_HAL_ChibiOS/I2CDevice.cpp

```diff
@@ -0,0 +1,219 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "I2CDevice.h"
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Math/AP_Math.h>
+#include "Util.h"
+#include "Scheduler.h"
+
+#include "ch.h"
+#include "hal.h"
+
+static I2CDriver* I2CD[] = { HAL_I2C_DEVICE_LIST };
+
+static uint8_t tx_dma_stream[] = {
+    STM32_I2C_I2C2_TX_DMA_STREAM,
+    STM32_I2C_I2C1_TX_DMA_STREAM
+};
+
+static uint8_t rx_dma_stream[] = {
+    STM32_I2C_I2C2_RX_DMA_STREAM,
+    STM32_I2C_I2C1_RX_DMA_STREAM
+};
+
+using namespace ChibiOS;
+extern const AP_HAL::HAL& hal;
+
+I2CBus I2CDeviceManager::businfo[ARRAY_SIZE(I2CD)];
+
+// get a handle for DMA sharing DMA channels with other subsystems
+void I2CBus::dma_init(void)
+{
+    dma_handle = new Shared_DMA(tx_dma_stream[busnum], rx_dma_stream[busnum],
+                                FUNCTOR_BIND_MEMBER(&I2CBus::dma_allocate, void),
+                                FUNCTOR_BIND_MEMBER(&I2CBus::dma_deallocate, void));    
+}
+
+// setup I2C buses
+I2CDeviceManager::I2CDeviceManager(void)
+{
+    for (uint8_t i=0; i<ARRAY_SIZE(I2CD); i++) {
+        businfo[i].busnum = i;
+        businfo[i].dma_init();
+        /*
+          setup default I2C config. As each device is opened we will
+          drop the speed to be the minimum speed requested
+         */
+        businfo[i].i2ccfg.op_mode = OPMODE_I2C;
+        businfo[i].i2ccfg.clock_speed = 400000;
+        businfo[i].i2ccfg.duty_cycle = FAST_DUTY_CYCLE_2;
+    }
+}
+
+I2CDevice::I2CDevice(uint8_t busnum, uint8_t address, uint32_t bus_clock, bool use_smbus, uint32_t timeout_ms) :
+    _retries(2),
+    _address(address),
+    _use_smbus(use_smbus),
+    _timeout_ms(timeout_ms),
+    bus(I2CDeviceManager::businfo[busnum])
+{
+    set_device_bus(busnum);
+    set_device_address(address);
+    asprintf(&pname, "I2C:%u:%02x",
+             (unsigned)busnum, (unsigned)address);
+    if (bus_clock < bus.i2ccfg.clock_speed) {
+        bus.i2ccfg.clock_speed = bus_clock;
+        if (bus_clock <= 100000) {
+            bus.i2ccfg.duty_cycle = STD_DUTY_CYCLE;
+        }
+    }
+}
+
+I2CDevice::~I2CDevice()
+{
+    printf("I2C device bus %u address 0x%02x closed\n", 
+           (unsigned)bus.busnum, (unsigned)_address);
+    free(pname);
```

Comment:
it shouldn't need to free dma_handle too ?

Reply:
dma_handle is at the bus level, not device level